src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -0,0 +1,753 @@
+                let existing_epoch = existing_region_info.region.get_region_epoch();
+                if epoch.get_version() <= existing_epoch.get_version()
+                    && epoch.get_conf_ver() <= existing_epoch.get_conf_ver()
+                {
+                    return;
+                }
+            }
+
+            let old_region = &mut existing_region_info.region;
+            assert_eq!(old_region.get_id(), region.get_id());
+
+            // If the end_key changed, the old item in `region_ranges` should be removed.
+            if old_region.get_end_key() != region.get_end_key() {
+                end_key_updated = true;
+                // The region's end_key has changed.
+                // Remove the old entry in `self.region_ranges`.
+                let old_end_key = data_end_key(old_region.get_end_key());
+
+                let old_id = self.region_ranges.remove(&old_end_key).unwrap();
+                assert_eq!(old_id, region.get_id());
+            }
+
+            // If the region already exists, update it and keep the original role.
+            *old_region = region.clone();
+        }
+
+        // The new region info has a different end_key, so we need to update it in the
+        // `region_ranges` map.
+        if end_key_updated {
+            // If the region already exists, then `region_ranges` need to be updated.
+            // However, if the new_end_key has already mapped to another region, we should only
+            // keep the one with higher `version`.
+            let end_key = data_end_key(region.get_end_key());
+            if !self.check_exist(
+                region.get_id(),
+                region.get_region_epoch().get_version(),
+                &end_key,
+            ) {
+                self.regions.remove(&region.get_id());
+                return;
+            }
+            // Insert new entry to `region_ranges`.
+            self.region_ranges.insert(end_key, region.get_id());
+        }
+    }
+
+    fn handle_create_region(&mut self, region: Region) {
+        // During tests, we found that the `Create` event may arrive multiple times. And when we
+        // receive an `Update` message, the region may have been deleted for some reason. So we
+        // handle it according to whether the region exists in the collection.
+        if self.regions.contains_key(&region.get_id()) {
+            info!("region_collection: trying to create region {} but it already exists, try to update it", region.get_id());
+            self.update_region(region);
+        } else {
+            self.create_region(region);
+        }
+    }
+
+    fn handle_update_region(&mut self, region: Region) {
+        if self.regions.contains_key(&region.get_id()) {
+            self.update_region(region);
+        } else {
+            info!("region_collection: trying to update region {} but it doesn't exist, try to create it", region.get_id());
+            self.create_region(region);
```

Comment:
Which role this region is? Leader or Follower?

Reply:
Here we simply set it follower. Since this case hardly happen (as shown in my tests), I think it's ok to fix it later.

Reply:
If this happens, GC will not be triggered for this region for a long time?

Reply:
How about always pass the `StateRole` in?